### ISSUE TITLE
BinaryQueue does not work correctly

### ISSUE THREAD

**Issue by int999** [RuedigerMoeller/fast-serialization#85]
The following code reproduces the problem : 

```Java
import org.nustaq.offheap.BinaryQueue;

public class TestQ {
    public static void main(String s[]) {		
		
		BinaryQueue bq = new BinaryQueue(3);
		bq.add((byte)1);
		bq.add((byte)2);
		bq.add((byte)3);
		bq.add((byte)4);
    			
		for (int i = 0; i < 4; i++) {
			System.out.println(bq.poll());
		}
	}
}

Output : 

4
-1
-1
-1

```

It seems the problem is that when the third byte is added, the code 

```
if ( addIndex >= storage.length() )
            addIndex -= storage.length();
```

in the BinaryQueue's add() method sets the addIndex to 0, and then when the fourth byte is added the remaining() method returns 3, when it should return 0.

**Comment by RuedigerMoeller** [comment]
oops, its in use at many places but with larger byte[] snippets. Thx for report

**Comment by RuedigerMoeller** [comment]
actually auto-grow should be triggered in your test case

**Comment by RuedigerMoeller** [comment]
it works for blobs, but not on individual bytes (probably unused), see https://github.com/RuedigerMoeller/fast-serialization/blob/master/src/test/ser/TestQ.java

byte wise r/w needs a fix :)

**Comment by RuedigerMoeller** [comment]
fixed with 2.40. The queue works as a ringbuffer. Pre-fix, an empty queue as well as a full queue had readIndex == writeIndex, which was discovered by testcase. Solution is to grow the q such that there is always at least one byte writable. note that backpressure must be applied by using the remaining() or capacity() method as else an outperforming writer will grow the queue uncontrolled.
Thanx for report, good finding.
